.github/workflows/e2e-tests.yml

```diff
@@ -44,171 +44,220 @@ jobs:
-        path: build-output/build/${{ github.event.client_payload.repository }}
-
-    - name: Prepare build
-      env:
-        COMPOSER_ROOT_VERSION: "dev-trunk"
-        BUILD_DIR: ./build-output
-        PROJECT_PATH: ${{ matrix.path }}
-        SUITE: ${{ matrix.suite }}
-      run: |
-        if [ "$GITHUB_EVENT_NAME" == workflow_run ]; then
-          echo "::group::Prepare build artefacts"
-          tar --xz -xf "$BUILD_DIR/jetpack-build/build.tar.xz" -C "$BUILD_DIR"
-          echo "::endgroup::"
-        fi
+      - name: Restore ${{ matrix.buildGroup }} build cache
+        id: jetpack-build-cache
+        uses: actions/cache/restore@v3
+        with:
+          path: .
+          key: ${{ matrix.buildGroup }}-${{ github.sha }}
+      - name: Checkout code
+        if: steps.jetpack-build-cache.outputs.cache-hit != 'true'
+        uses: actions/checkout@v3
 
-        if [ "$GITHUB_EVENT_NAME" == workflow_run ] || [ "$GITHUB_EVENT_NAME" == repository_dispatch ]; then
-          echo "::group::Update volume mapping"
-          .github/files/e2e-tests/map-plugins-for-e2e-env.sh
-          echo "::endgroup::"
-        else
+      - name: Setup tools
+        if: steps.jetpack-build-cache.outputs.cache-hit != 'true'
+        uses: ./.github/actions/tool-setup
+
+      - name: Install monorepo
+        if: steps.jetpack-build-cache.outputs.cache-hit != 'true'
+        run: pnpm install
+
+      - name: Build projects
+        if: steps.jetpack-build-cache.outputs.cache-hit != 'true'
+        env:
+          COMPOSER_ROOT_VERSION: "dev-trunk"
+          BUILD_DIR: ./build-output
+          PROJECT_PATH: ${{ matrix.path }}
+          SUITE: ${{ matrix.suite }}
+        run: |
           if [[ "${SUITE}" != atomic* ]]; then
             echo "::group::Build plugin(s)"
             cd $PROJECT_PATH
             pnpm run build
             echo "::endgroup::"
           fi
-        fi
-
-    - name: Test environment set-up
-      working-directory: ${{ matrix.path }}
-      env:
-        CONFIG_KEY: ${{ secrets.E2E_CONFIG_KEY }}
-        SUITE: ${{ matrix.suite }}
-        PROJECT_NAME:  ${{ matrix.project }}
-      run: |
-        echo "::group::Decrypt config"
-        pnpm run config:decrypt
-        echo "::endgroup::"
-
-        if [[ "${SUITE}" != atomic* && "${SUITE}" != vip ]]; then
-          echo "::group::Start docker environment"
-          pnpm run env:up
-          echo "::endgroup::"
+      - name: Save ${{ matrix.buildGroup }} build cache
+        if: steps.jetpack-build-cache.outputs.cache-hit != 'true'
+        id: jetpack-build-cache-save
+        uses: actions/cache/save@v3
+        with:
+          path: .
+          key: ${{ steps.jetpack-build-cache.outputs.cache-primary-key }}
+
+  e2e-tests:
+    name: "${{ matrix.project }} e2e tests"
+    runs-on: ubuntu-latest
+    if: ${{ always() }} &&  github.event.pull_request.head.repo.full_name == github.event.pull_request.base.repo.full_name
```

Comment:
I probably don't, but "skips" are considered failures - because of the concurrency statement above - the e2e-tests never start after the build is done without `always()`

Reply:
As written you won't ever have any skips I think. Although you probably will once you fix some of the other comments. OTOH, you will have cancellations, which I recommended you fix above (line 58).

If https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions is accurate, skips should fall under the default `success()` as they're not failure or cancel. But if they don't, you'd probably want something along the lines of `always() && ( needs.e2e-tests.result == 'success' || needs.e2e-tests.result == 'skipped' )` to still only run on success or skip.